sklearn/covariance/tests/test_covariance.py

```diff
@@ -57,10 +57,10 @@ def test_covariance():
         cov.error_norm(empirical_covariance(X_1d), norm='spectral'), 0)
 
     # test with one sample
-    X_1sample = np.arange(5)
-    cov = EmpiricalCovariance()
-
-    assert_warns(UserWarning, cov.fit, X_1sample)
+    # FIXME I don't know what this test does
+    # X_1sample = np.arange(5)
+    # cov = EmpiricalCovariance()
```

Comment:
In master I get the following:

```
>>> emp_cov = EmpiricalCovariance().fit(np.arange(5))
/Users/ogrisel/code/scikit-learn/sklearn/covariance/empirical_covariance_.py:73: UserWarning: Only one sample available. You may want to reshape your data array
  warnings.warn("Only one sample available. "
>>> emp_cov.covariance_
array([[ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.]])
>>> emp_cov.precision_
array([[ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.]])
```

For regularized models:

```python
>>> lw_cov = LedoitWolf().fit(np.arange(5))
/Users/ogrisel/code/scikit-learn/sklearn/covariance/shrunk_covariance_.py:283: UserWarning: Only one sample available. You may want to reshape your data array
  warnings.warn("Only one sample available. "
>>> lw_cov.covariance_
array([[ 4.,  2.,  0., -2., -4.],
       [ 2.,  1.,  0., -1., -2.],
       [ 0.,  0.,  0.,  0.,  0.],
       [-2., -1.,  0.,  1.,  2.],
       [-4., -2.,  0.,  2.,  4.]])
>>> lw_cov.precision_
array([[ 0.04,  0.02,  0.  , -0.02, -0.04],
       [ 0.02,  0.01,  0.  , -0.01, -0.02],
       [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
       [-0.02, -0.01,  0.  ,  0.01,  0.02],
       [-0.04, -0.02,  0.  ,  0.02,  0.04]])
```

So indeed covariance mode seems to treat the case of 1D input as 1 sample cases which is not consistent with the rest of the library. On this branch I get:

```
>>> emp_cov = EmpiricalCovariance().fit(np.arange(5))
>>> emp_cov.covariance_
array([[ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.]])
```

So no more warning but still the covariance matrix is 5x5 five meaning that the 1D input is still treated as a single sample dataset which is still not consistent with the rest of the library IIRC.


Anyway covariance models of single sample datasets are pretty much meaningless to me. Covariance models of datasets with a single features are barely better: they are singleton matrices which while well defined (this is just the variance of the feature), is pretty useless to treat as a covariance estimation model in practice.

I think we should consider the case of covariance estimation of datasets with a single sample as invalid and raise a `ValueError` telling explicitly that X needs at least 2 samples.